File: scala-fit-performer/src/main/scala/com/couchbase/client/performer/scala/manager/BucketManagerHelper.scala
/*
 * Copyright (c) 2023 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.performer.scala.manager

import com.couchbase.client.performer.scala.ScalaSdkCommandExecutor.{convertDurability, setSuccess}
import com.couchbase.client.performer.scala.util.OptionsUtil.{
  DefaultManagementTimeout,
  convertDuration
}
import com.couchbase.client.protocol.run.Result
import com.couchbase.client.protocol.sdk.Command
import com.couchbase.client.protocol.shared.Durability
import com.couchbase.client.scala.durability.Durability._
import com.couchbase.client.scala.manager.bucket
import com.couchbase.client.scala.manager.bucket.BucketType.{Couchbase, Ephemeral, Memcached}
import com.couchbase.client.scala.manager.bucket.CompressionMode.{Active, Off}
import com.couchbase.client.scala.manager.bucket.{
  BucketSettings,
  CompressionMode,
  ConflictResolutionType,
  CreateBucketSettings,
  EjectionMethod,
  StorageBackend
}
import com.couchbase.client.scala.{Cluster, ReactiveCluster}
import com.google.protobuf.Duration
import reactor.core.scala.publisher.SMono

import java.util.concurrent.TimeUnit
import scala.concurrent.duration.Duration
import scala.util.{Failure, Success}
import scala.jdk.CollectionConverters._

object BucketManagerHelper {

  def handleBucketManager(cluster: Cluster, command: Command): Result.Builder = {
    val bm = command.getClusterCommand.getBucketManager

    val result = Result.newBuilder()

    if (bm.hasGetBucket) {
      val bucketName = bm.getGetBucket.getBucketName

      val response = cluster.buckets.getBucket(
        bucketName,
        if (bm.getGetBucket.hasOptions && bm.getGetBucket.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(bucketSettings) => populateResult(result, bucketSettings)
        case Failure(e)              => throw e
      }
    } else if (bm.hasGetAllBuckets) {
      val req = bm.getGetAllBuckets
      val response = cluster.buckets.getAllBuckets(
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(buckets) => populateResult(result, buckets)
        case Failure(e)       => throw e
      }
    } else if (bm.hasCreateBucket) {
      val req = bm.getCreateBucket
      val response = cluster.buckets.create(
        createBucketSettings(req.getSettings.getSettings, Some(req.getSettings)),
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasDropBucket) {
      val req = bm.getDropBucket
      val response = cluster.buckets.dropBucket(
        req.getBucketName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasFlushBucket) {
      val req = bm.getFlushBucket
      val response = cluster.buckets.flushBucket(
        req.getBucketName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else if (bm.hasUpdateBucket) {
      val req = bm.getUpdateBucket
      val response = cluster.buckets.updateBucket(
        createBucketSettings(req.getSettings, None),
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(req.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response match {
        case Success(_) => setSuccess(result)
        case Failure(e) => throw e
      }
    } else {
      throw new UnsupportedOperationException(new IllegalArgumentException("Unknown operation"))
    }

    result
  }

  def handleBucketManagerReactive(cluster: ReactiveCluster, command: Command): SMono[Result] = {
    val bm = command.getClusterCommand.getBucketManager

    val result = Result.newBuilder()

    if (bm.hasGetBucket) {
      val bucketName = bm.getGetBucket.getBucketName

      val response: SMono[bucket.BucketSettings] = cluster.buckets.getBucket(
        bucketName,
        if (bm.getGetBucket.hasOptions && bm.getGetBucket.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.map(r => {
        populateResult(result, r)
        result.build()
      })
    } else if (bm.hasGetAllBuckets) {
      val req = bm.getGetAllBuckets
      val response = cluster.buckets.getAllBuckets(
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.collectSeq.map(buckets => populateResult(result, buckets).build)
    } else if (bm.hasCreateBucket) {
      val req = bm.getCreateBucket
      val response = cluster.buckets.create(
        createBucketSettings(req.getSettings.getSettings, Some(req.getSettings)),
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.map(_ => {
        setSuccess(result)
        result.build
      })
    } else if (bm.hasDropBucket) {
      val req = bm.getDropBucket
      val response = cluster.buckets.dropBucket(
        req.getBucketName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.map(_ => {
        setSuccess(result)
        result.build
      })
    } else if (bm.hasFlushBucket) {
      val req = bm.getFlushBucket
      val response = cluster.buckets.flushBucket(
        req.getBucketName,
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.map(_ => {
        setSuccess(result)
        result.build
      })
    } else if (bm.hasUpdateBucket) {
      val req = bm.getUpdateBucket
      val response = cluster.buckets.updateBucket(
        createBucketSettings(req.getSettings, None),
        if (req.hasOptions && req.getOptions.hasTimeoutMsecs)
          scala.concurrent.duration
            .Duration(bm.getGetBucket.getOptions.getTimeoutMsecs, TimeUnit.MILLISECONDS)
        else DefaultManagementTimeout
      )

      response.map(_ => {
        setSuccess(result)
        result.build
      })
    } else {
      SMono.error(new IllegalArgumentException("Unknown operation"))
    }
  }

  def createBucketSettings(
      bs: com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketSettings,
      cbs: Option[com.couchbase.client.protocol.sdk.cluster.bucketmanager.CreateBucketSettings]
  ): CreateBucketSettings = {
    var cs = CreateBucketSettings(bs.getName, bs.getRamQuotaMB.toInt)
    if (bs.hasFlushEnabled) cs = cs.flushEnabled(bs.getFlushEnabled)
    if (bs.hasNumReplicas) cs = cs.numReplicas(bs.getNumReplicas)
    if (bs.hasReplicaIndexes) cs = cs.replicaIndexes(bs.getReplicaIndexes)
    if (bs.hasBucketType) cs = cs.bucketType(bs.getBucketType match {
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.COUCHBASE =>
        bucket.BucketType.Couchbase
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.EPHEMERAL =>
        bucket.BucketType.Ephemeral
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.MEMCACHED =>
        bucket.BucketType.Memcached
      case _ => throw new UnsupportedOperationException("Unknown bucket type")
    })
    if (bs.hasEvictionPolicy) cs = cs.ejectionMethod(bs.getEvictionPolicy match {
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.FULL =>
        EjectionMethod.FullEviction
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.NO_EVICTION =>
        EjectionMethod.NoEviction
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.NOT_RECENTLY_USED =>
        EjectionMethod.NotRecentlyUsed
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.VALUE_ONLY =>
        EjectionMethod.ValueOnly
      case _ => throw new UnsupportedOperationException("Unknown eviction policy type")
    })
    if (bs.hasMaxExpirySeconds) cs = cs.maxTTL(bs.getMaxExpirySeconds)
    if (bs.hasCompressionMode) cs = cs.compressionMode(bs.getCompressionMode match {
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.ACTIVE =>
        CompressionMode.Active
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.OFF =>
        CompressionMode.Off
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.PASSIVE =>
        CompressionMode.Passive
      case _ => throw new UnsupportedOperationException("Unknown compression mode")
    })
    if (bs.hasMinimumDurabilityLevel)
      cs = cs.minimumDurabilityLevel(convertDurability(bs.getMinimumDurabilityLevel))
    if (bs.hasStorageBackend) cs = cs.storageBackend(bs.getStorageBackend match {
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.StorageBackend.COUCHSTORE =>
        StorageBackend.Couchstore
      case com.couchbase.client.protocol.sdk.cluster.bucketmanager.StorageBackend.MAGMA =>
        StorageBackend.Magma
      case _ => throw new UnsupportedOperationException("Unknown storage backend")
    })
    cbs match {
      case Some(value) =>
        if (value.hasConflictResolutionType)
          cs = cs.conflictResolutionType(value.getConflictResolutionType match {
            case com.couchbase.client.protocol.sdk.cluster.bucketmanager.ConflictResolutionType.TIMESTAMP =>
              ConflictResolutionType.Timestamp
            case com.couchbase.client.protocol.sdk.cluster.bucketmanager.ConflictResolutionType.SEQUENCE_NUMBER =>
              ConflictResolutionType.SequenceNumber
            case com.couchbase.client.protocol.sdk.cluster.bucketmanager.ConflictResolutionType.CUSTOM =>
              ConflictResolutionType.Custom
            case _ => throw new UnsupportedOperationException("Unknown conflict resolution type")
          })
      case _ =>
    }
    cs
  }

  def populateResult(
      result: Result.Builder,
      response: com.couchbase.client.scala.manager.bucket.BucketSettings
  ): Result.Builder = {
    result.setSdk(
      com.couchbase.client.protocol.sdk.Result
        .newBuilder()
        .setBucketManagerResult(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.Result
            .newBuilder()
            .setBucketSettings(populateBucketSettings(response))
        )
    )
  }

  private def populateBucketSettings(
      response: bucket.BucketSettings
  ): com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketSettings = {
    val builder =
      com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketSettings.newBuilder()

    builder
      .setName(response.name)
      .setFlushEnabled(response.flushEnabled)
      .setRamQuotaMB(response.ramQuotaMB)
      .setNumReplicas(response.numReplicas)
      .setReplicaIndexes(response.replicaIndexes)
      .setMaxExpirySeconds(response.maxTTL.getOrElse(0))

    response.bucketType match {
      case Couchbase =>
        builder.setBucketType(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.COUCHBASE
        )
      case Memcached =>
        builder.setBucketType(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.MEMCACHED
        )
      case Ephemeral =>
        builder.setBucketType(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType.EPHEMERAL
        )
    }

    response.ejectionMethod match {
      case EjectionMethod.FullEviction =>
        builder.setEvictionPolicy(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.FULL
        )
      case EjectionMethod.NoEviction =>
        builder.setEvictionPolicy(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.NO_EVICTION
        )
      case EjectionMethod.NotRecentlyUsed =>
        builder.setEvictionPolicy(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.NOT_RECENTLY_USED
        )
      case EjectionMethod.ValueOnly =>
        builder.setEvictionPolicy(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType.VALUE_ONLY
        )
    }

    response.compressionMode.orNull match {
      case Off =>
        builder.setCompressionMode(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.OFF
        )
      case Active =>
        builder.setCompressionMode(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.ACTIVE
        )
      case _ =>
        builder.setCompressionMode(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode.PASSIVE
        )
    }

    response.minimumDurabilityLevel match {
      case Disabled             => builder.setMinimumDurabilityLevel(Durability.NONE)
      case ClientVerified(_, _) => builder.setMinimumDurabilityLevel(Durability.NONE)
      case Majority             => builder.setMinimumDurabilityLevel(Durability.MAJORITY)
      case MajorityAndPersistToActive =>
        builder.setMinimumDurabilityLevel(Durability.MAJORITY_AND_PERSIST_TO_ACTIVE)
      case PersistToMajority => builder.setMinimumDurabilityLevel(Durability.PERSIST_TO_MAJORITY)
    }

    builder.build
  }

  def populateResult(
      result: Result.Builder,
      response: Seq[com.couchbase.client.scala.manager.bucket.BucketSettings]
  ): Result.Builder = {

    result.setSdk(
      com.couchbase.client.protocol.sdk.Result
        .newBuilder()
        .setBucketManagerResult(
          com.couchbase.client.protocol.sdk.cluster.bucketmanager.Result
            .newBuilder()
            .setGetAllBucketsResult(
              com.couchbase.client.protocol.sdk.cluster.bucketmanager.GetAllBucketsResult.newBuilder
                .putAllResult(
                  response.map(x => x.name -> populateBucketSettings(x)).toMap.asJava
                )
            )
        )
    )
  }
}

File: kotlin-fit-performer/src/main/kotlin/com/couchbase/client/performer/kotlin/manager/BucketManagerHelper.kt
/*
 * Copyright (c) 2023 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.performer.kotlin.manager

import com.couchbase.client.kotlin.Cluster
import com.couchbase.client.kotlin.CommonOptions
import com.couchbase.client.kotlin.kv.Durability
import com.couchbase.client.kotlin.kv.Expiry
import com.couchbase.client.kotlin.manager.bucket.BucketSettings
import com.couchbase.client.kotlin.manager.bucket.BucketType.Companion.COUCHBASE
import com.couchbase.client.kotlin.manager.bucket.BucketType.Companion.EPHEMERAL
import com.couchbase.client.kotlin.manager.bucket.BucketType.Companion.MEMCACHED
import com.couchbase.client.kotlin.manager.bucket.CompressionMode
import com.couchbase.client.kotlin.manager.bucket.ConflictResolutionType
import com.couchbase.client.kotlin.manager.bucket.EvictionPolicyType.Companion.FULL
import com.couchbase.client.kotlin.manager.bucket.EvictionPolicyType.Companion.NOT_RECENTLY_USED
import com.couchbase.client.kotlin.manager.bucket.EvictionPolicyType.Companion.NO_EVICTION
import com.couchbase.client.kotlin.manager.bucket.EvictionPolicyType.Companion.VALUE_ONLY
import com.couchbase.client.kotlin.manager.bucket.StorageBackend
import com.couchbase.client.kotlin.manager.bucket.StorageBackend.Companion.COUCHSTORE
import com.couchbase.client.kotlin.manager.bucket.StorageBackend.Companion.MAGMA
import com.couchbase.client.kotlin.util.StorageSize.Companion.bytes
import com.couchbase.client.kotlin.util.StorageSize.Companion.mebibytes
import com.couchbase.client.protocol.run.Result
import com.couchbase.client.protocol.sdk.Command
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.CreateBucketRequest
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.DropBucketRequest
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.FlushBucketRequest
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.GetAllBucketsRequest
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.GetAllBucketsResult
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.GetBucketOptions
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.GetBucketRequest
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.UpdateBucketRequest
import kotlin.time.Duration.Companion.seconds
import kotlin.time.DurationUnit
import kotlin.time.toDuration
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketSettings as FitBucketSettings
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.BucketType as FitBucketType
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.CompressionMode as FitCompressionMode
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.ConflictResolutionType as FitConflictResolutionType
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.EvictionPolicyType as FitEvictionPolicyType
import com.couchbase.client.protocol.sdk.cluster.bucketmanager.StorageBackend as FitStorageBackend
import com.couchbase.client.protocol.shared.Durability as FitDurability

private fun unrecognized(type: String, value: Any?): Nothing {
    throw RuntimeException("Unrecognized $type: $value")
}

suspend fun dropBucket(cluster: Cluster, request: DropBucketRequest, result: Result.Builder) {
    cluster.buckets.dropBucket(request.bucketName)
    result.success()
}

suspend fun flushBucket(cluster: Cluster, request: FlushBucketRequest, result: Result.Builder) {
    cluster.buckets.flushBucket(request.bucketName)
    result.success()
}

suspend fun createBucket(cluster: Cluster, request: CreateBucketRequest, result: Result.Builder) {
    val settings = request.settings.settings

    cluster.buckets.createBucket(
        name = settings.name,
        common = CommonOptions.Default, // todo
        ramQuota = settings.ramQuotaMB.mebibytes,
        bucketType = if (!settings.hasBucketType()) null else when (settings.bucketType) {
            FitBucketType.COUCHBASE -> COUCHBASE
            FitBucketType.EPHEMERAL -> EPHEMERAL
            FitBucketType.MEMCACHED -> MEMCACHED
            else -> unrecognized("bucket type", settings.bucketType)
        },
        storageBackend = if (!settings.hasStorageBackend()) null else when (settings.storageBackend) {
            FitStorageBackend.COUCHSTORE -> COUCHSTORE
            FitStorageBackend.MAGMA -> MAGMA
            else -> unrecognized("storage backend", settings.storageBackend)
        },
        evictionPolicy = if (!settings.hasEvictionPolicy()) null else when (settings.evictionPolicy) {
            FitEvictionPolicyType.FULL -> FULL
            FitEvictionPolicyType.NOT_RECENTLY_USED -> NOT_RECENTLY_USED
            FitEvictionPolicyType.NO_EVICTION -> NO_EVICTION
            FitEvictionPolicyType.VALUE_ONLY -> VALUE_ONLY
            else -> unrecognized("eviction policy", settings.evictionPolicy)
        },
        flushEnabled = if (!settings.hasFlushEnabled()) null else settings.flushEnabled,
        replicas = if (!settings.hasNumReplicas()) null else settings.numReplicas,
        maximumExpiry = if (!settings.hasMaxExpirySeconds()) null else Expiry.of(settings.maxExpirySeconds.seconds),
        compressionMode = if (!settings.hasCompressionMode()) null else when (settings.compressionMode) {
            FitCompressionMode.ACTIVE -> CompressionMode.ACTIVE
            FitCompressionMode.OFF -> CompressionMode.Companion.OFF
            FitCompressionMode.PASSIVE -> CompressionMode.PASSIVE
            else -> unrecognized("compression mode", settings.compressionMode)
        },
        minimumDurability = if (!settings.hasMinimumDurabilityLevel()) null else when (settings.minimumDurabilityLevel) {
            FitDurability.NONE -> Durability.none()
            FitDurability.MAJORITY -> Durability.majority()
            FitDurability.MAJORITY_AND_PERSIST_TO_ACTIVE -> Durability.majorityAndPersistToActive()
            FitDurability.PERSIST_TO_MAJORITY -> Durability.persistToMajority()
            else -> unrecognized("durability", settings.minimumDurabilityLevel)
        },
        conflictResolutionType = if (!request.settings.hasConflictResolutionType()) null else when (request.settings.conflictResolutionType) {
            FitConflictResolutionType.TIMESTAMP -> ConflictResolutionType.TIMESTAMP
            FitConflictResolutionType.SEQUENCE_NUMBER -> ConflictResolutionType.SEQUENCE_NUMBER
            FitConflictResolutionType.CUSTOM -> ConflictResolutionType.CUSTOM
            else -> unrecognized("conflict resolution type", request.settings.conflictResolutionType)
        },
        replicateViewIndexes = if (!settings.hasReplicaIndexes()) null else settings.replicaIndexes,
        // [start:1.2.0]
        historyRetentionCollectionDefault = if (!settings.hasHistoryRetentionCollectionDefault()) null else settings.historyRetentionCollectionDefault,
        historyRetentionSize = if (!settings.hasHistoryRetentionBytes()) null else settings.historyRetentionBytes.bytes,
        historyRetentionDuration = if (!settings.hasHistoryRetentionSeconds()) null else settings.historyRetentionSeconds.seconds,
        // [end:1.2.0]
    )

    result.success()
}

suspend fun updateBucket(cluster: Cluster, request: UpdateBucketRequest, result: Result.Builder) {
    val settings = request.settings

    cluster.buckets.updateBucket(
        name = settings.name,
        common = CommonOptions.Default, // todo
        ramQuota = settings.ramQuotaMB.let { if (it == 0L) null else it.mebibytes },
        evictionPolicy = if (!settings.hasEvictionPolicy()) null else when (settings.evictionPolicy) {
            FitEvictionPolicyType.FULL -> FULL
            FitEvictionPolicyType.NOT_RECENTLY_USED -> NOT_RECENTLY_USED
            FitEvictionPolicyType.NO_EVICTION -> NO_EVICTION
            FitEvictionPolicyType.VALUE_ONLY -> VALUE_ONLY
            else -> unrecognized("eviction policy", settings.evictionPolicy)
        },
        flushEnabled = if (!settings.hasFlushEnabled()) null else settings.flushEnabled,
        replicas = if (!settings.hasNumReplicas()) null else settings.numReplicas,
        maximumExpiry = if (!settings.hasMaxExpirySeconds()) null else {
            if (settings.maxExpirySeconds == 0) Expiry.none() else Expiry.of(settings.maxExpirySeconds.seconds)
        },
        compressionMode = if (!settings.hasCompressionMode()) null else when (settings.compressionMode) {
            FitCompressionMode.ACTIVE -> CompressionMode.ACTIVE
            FitCompressionMode.OFF -> CompressionMode.Companion.OFF
            FitCompressionMode.PASSIVE -> CompressionMode.PASSIVE
            else -> unrecognized("compression mode", settings.compressionMode)
        },
        minimumDurability = if (!settings.hasMinimumDurabilityLevel()) null else when (settings.minimumDurabilityLevel) {
            FitDurability.NONE -> Durability.none()
            FitDurability.MAJORITY -> Durability.majority()
            FitDurability.MAJORITY_AND_PERSIST_TO_ACTIVE -> Durability.majorityAndPersistToActive()
            FitDurability.PERSIST_TO_MAJORITY -> Durability.persistToMajority()
            else -> unrecognized("durability", settings.minimumDurabilityLevel)
        },
        // [start:1.2.0]
        historyRetentionCollectionDefault = if (!settings.hasHistoryRetentionCollectionDefault()) null else settings.historyRetentionCollectionDefault,
        historyRetentionSize = if (!settings.hasHistoryRetentionBytes()) null else settings.historyRetentionBytes.bytes,
        historyRetentionDuration = if (!settings.hasHistoryRetentionSeconds()) null else settings.historyRetentionSeconds.seconds,
        // [end:1.2.0]
    )

    result.success()
}

suspend fun handleBucketManager(cluster: Cluster, command: Command, result: Result.Builder) {
    val bm = command.clusterCommand.bucketManager
    when {
        bm.hasGetBucket() -> getBucket(cluster, bm.getBucket, result)
        bm.hasGetAllBuckets() -> getAllBucket(cluster, bm.getAllBuckets, result)
        bm.hasCreateBucket() -> createBucket(cluster, bm.createBucket, result)
        bm.hasUpdateBucket() -> updateBucket(cluster, bm.updateBucket, result)
        bm.hasDropBucket() -> dropBucket(cluster, bm.dropBucket, result)
        bm.hasFlushBucket() -> flushBucket(cluster, bm.flushBucket, result)
    }
}

suspend fun getBucket(cluster: Cluster, request: GetBucketRequest, result: Result.Builder) {
    val bucketName = request.bucketName
    val response: BucketSettings
    if (request.hasOptions()) {
        response = cluster.buckets.getBucket(bucketName)
    } else {
        val options = createGetBucketOptions(request.options)
        response = cluster.buckets.getBucket(bucketName, options)
    }
    populateResult(result, response)
}

suspend fun getAllBucket(cluster: Cluster, request: GetAllBucketsRequest, result: Result.Builder) {
    val s: Map<String, FitBucketSettings> = cluster.buckets.getAllBuckets()
        .groupBy(BucketSettings::name) { it.toFit() }
        .mapValues { it.value.single().build() }

    result.setSdk(
        com.couchbase.client.protocol.sdk.Result.newBuilder()
            .setBucketManagerResult(
                com.couchbase.client.protocol.sdk.cluster.bucketmanager.Result.newBuilder()
                    .setGetAllBucketsResult(GetAllBucketsResult.newBuilder().putAllResult(s))
            )
    )
}

fun createGetBucketOptions(getBucketOptions: GetBucketOptions): CommonOptions {
    val options = if (getBucketOptions.hasTimeoutMsecs()) {
        CommonOptions(getBucketOptions.timeoutMsecs.toDuration(DurationUnit.MILLISECONDS))
    } else {
        CommonOptions.Default
    }
    return options
}

private fun BucketSettings.toFit(): FitBucketSettings.Builder {
    val sdk = this
    return FitBucketSettings.newBuilder().apply {
        compressionMode = FitCompressionMode.valueOf(sdk.compressionMode.toString().uppercase())
        flushEnabled = sdk.flushEnabled
        name = sdk.name
        numReplicas = sdk.replicas
        ramQuotaMB = sdk.ramQuota.inBytes / 1.mebibytes.inBytes

        bucketType = when (sdk.bucketType) {
            COUCHBASE -> FitBucketType.COUCHBASE
            EPHEMERAL -> FitBucketType.EPHEMERAL
            MEMCACHED -> FitBucketType.MEMCACHED
            else -> unrecognized("bucket type", sdk.bucketType)
        }

        evictionPolicy = when (sdk.evictionPolicy) {
            FULL -> FitEvictionPolicyType.FULL
            NO_EVICTION -> FitEvictionPolicyType.NO_EVICTION
            NOT_RECENTLY_USED -> FitEvictionPolicyType.NOT_RECENTLY_USED
            VALUE_ONLY -> FitEvictionPolicyType.VALUE_ONLY
            else -> unrecognized("eviction policy", sdk.evictionPolicy)
        }

        minimumDurabilityLevel = when (sdk.minimumDurability) {
            Durability.majority() -> FitDurability.MAJORITY
            Durability.majorityAndPersistToActive() -> FitDurability.MAJORITY_AND_PERSIST_TO_ACTIVE
            Durability.persistToMajority() -> FitDurability.PERSIST_TO_MAJORITY
            else -> FitDurability.NONE
        }

        when (sdk.storageBackend) {
            COUCHSTORE -> storageBackend = FitStorageBackend.COUCHSTORE
            MAGMA -> storageBackend = FitStorageBackend.MAGMA
            StorageBackend.of("undefined") -> {} // don't set; ephemeral / memcached
            StorageBackend.of("") -> {} // don't set; old server doesn't know about storage backends
            else -> unrecognized("storage backend", sdk.storageBackend)
        }

        maxExpirySeconds = when (sdk.maximumExpiry) {
            Expiry.None -> 0
            Expiry.Unknown -> 0
            else -> (sdk.maximumExpiry as Expiry.Relative).duration.inWholeSeconds.toInt()
        }

        sdk.historyRetentionCollectionDefault?.let { historyRetentionCollectionDefault = it }
        sdk.historyRetentionDuration?.let { historyRetentionSeconds = it.inWholeSeconds }
        sdk.historyRetentionSize?.let { historyRetentionBytes = it.inBytes }
    }
}

private fun populateResult(result: Result.Builder, response: BucketSettings) {
    val builder = response.toFit()

    result.setSdk(
        com.couchbase.client.protocol.sdk.Result.newBuilder()
            .setBucketManagerResult(
                com.couchbase.client.protocol.sdk.cluster.bucketmanager.Result.newBuilder()
                    .setBucketSettings(builder)
            )
    )
}

private fun Result.Builder.success() {
    setSdk(com.couchbase.client.protocol.sdk.Result.newBuilder().setSuccess(true))
}
